src/SignalR/clients/cpp/samples/HubConnectionSample/HubConnectionSample.cpp

```diff
@@ -41,44 +45,63 @@ void send_message(signalr::hub_connection& connection, const std::string& messag
 void chat()
 {
     signalr::hub_connection connection("http://localhost:5000/default", signalr::trace_level::all, std::make_shared<logger>());
-    connection.on("Send", [](const web::json::value& m)
+    connection.on("Send", [](const web::json::value & m)
     {
         ucout << std::endl << m.at(0).as_string() << /*U(" wrote:") << m.at(1).as_string() <<*/ std::endl << U("Enter your message: ");
     });
 
-    connection.start()
-        .then([&connection]()
+    std::promise<void> task;
+    connection.start([&connection, &task](std::exception_ptr exception)
+    {
+        if (exception)
         {
-            ucout << U("Enter your message:");
-            for (;;)
+            try
             {
-                std::string message;
-                std::getline(std::cin, message);
-
-                if (message == ":q")
-                {
-                    break;
-                }
 
-                send_message(connection, message);
             }
-        })
-        .then([&connection]() // fine to capture by reference - we are blocking so it is guaranteed to be valid
+            catch (const std::exception & ex)
+            {
+                ucout << U("exception when starting connection: ") << ex.what() << std::endl;
+            }
+            task.set_value();
+            return;
+        }
+
+        ucout << U("Enter your message:");
+        for (;;)
         {
-            return connection.stop();
-        })
-        .then([](pplx::task<void> stop_task)
+            std::string message;
+            std::getline(std::cin, message);
+
+            if (message == ":q")
+            {
+                break;
+            }
+
+            send_message(connection, message);
+        }
+
+        connection.stop([&task](std::exception_ptr exception)
         {
             try
             {
-                stop_task.get();
+                if (exception)
+                {
+                    std::rethrow_exception(exception);
+                }
+
                 ucout << U("connection stopped successfully") << std::endl;
             }
-            catch (const std::exception &e)
+            catch (const std::exception & e)
             {
-                ucout << U("exception when starting or stopping connection: ") << e.what() << std::endl;
+                ucout << U("exception when stopping connection: ") << e.what() << std::endl;
             }
-        }).get();
+
+            task.set_value();
```

Comment:
Wasn't this already completed in the start callback?

Reply:
The start callback wraps everything, and when we complete this task we also exit the callback. So it will never be set twice.